tower/src/util/and_then.rs

```diff
@@ -0,0 +1,74 @@
+use std::task::{Context, Poll};
+
+use futures_core::TryFuture;
+use futures_util::{
+    future::{AndThen as AndThenFut, ErrInto as ErrIntoFut},
+    TryFutureExt,
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Service returned by the [`and_then`] combinator.
+///
+/// [`and_then`]: crate::util::ServiceExt::and_then
+#[derive(Clone, Debug)]
+pub struct AndThen<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> AndThen<S, F> {
+    /// Creates a new `AndThen` service.
+    pub fn new(inner: S, f: F) -> Self {
+        AndThen { f, inner }
+    }
+}
+
+impl<S, F, Request, Error, Fut> Service<Request> for AndThen<S, F>
+where
+    S: Service<Request>,
+    S::Error: Into<Error>,
+    F: FnOnce(S::Response) -> Fut + Clone,
+    Fut: TryFuture<Error = Error>,
+{
+    type Response = Fut::Ok;
+    type Error = Error;
+    type Future = AndThenFut<ErrIntoFut<S::Future, Error>, Fut, F>;
```

Comment:
I think some of the generic parameters aren't necessary here:
```suggestion
impl<S, F, Request, Fut> Service<Request> for AndThen<S, F>
where
    S: Service<Request>,
    S::Error: Into<Fut::Error>,
    F: FnOnce(S::Response) -> Fut + Clone,
    Fut: TryFuture,
{
    type Response = Fut::Ok;
    type Error = Fut::Error;
    type Future = AndThenFut<ErrIntoFut<S::Future, Fut::Error>, Fut, F>;
```

Reply:
I will get an unnecessary parameter passed you eventually...